### PR TITLE
feat(vite-plugin-angular): add inline markdown support to analog template

### PR DESCRIPTION
**Pull request by brandonroberts**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/analogjs/analog/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [x] vite-plugin-angular
- [ ] vite-plugin-nitro
- [ ] astro-angular
- [ ] create-analog
- [ ] router
- [ ] platform
- [ ] content
- [ ] nx-plugin
- [ ] trpc

## What is the current behavior?

Closes #

## What is the new behavior?

An inline template for a `.analog` file supports authoring with markdown. This template is compiled into HTML and used as the component template, allowing for Angular syntax and components to be used.

```html
<script lang="ts">
  import Hello from '../hello.analog';
</script>

<template lang="md">
  # About me works!

  ## Subheader

  '''sh
  npm create analog
  '''

  <Hello />
</template>
```

Result:

image

NOTE: The same constraints exist for an Angular component template, such as using `@` directly, interpolation tags, and more.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by goetzrobin — COMMENTED** [review]
Love love love this PR! Just so I understand what is going on. If you use the `<template lang="md">` tag, the template get's turned into a virtual file, which is recognized by it's "virtual" path and parsed through the `marked` service while it is read. Which means the template contains "regular" html. Importantly this "regular" html can contain Angular elements, because it will then be again parsed through the default `.analog` readFile function, which means we end up with markdown that supports Angular components. Is that about right? 😍

File: packages/vite-plugin-angular/src/lib/host.ts [hunk 1]
```diff
@@ -53,16 +54,60 @@ export function augmentHostWithResources(
         ...parameters
       );
     };
+
+    const baseReadFile = (resourceHost as ts.CompilerHost).readFile;
+    const TEMPLATE_TAG_REGEX =
+      /(<template>|<template lang="md">)([\s\S]*?)<\/template>/i;
```
Comment (nartc) [inline packages/vite-plugin-angular/src/lib/host.ts:60]: suggestion: let's make these constants in a separate file

File: packages/vite-plugin-angular/src/lib/authoring/analog.ts [hunk 2]
```diff
@@ -77,7 +78,12 @@ export function compileAnalogFile(
   const componentMetadata = (() => {
     if (ngType === 'Component') {
       const items = ['changeDetection: ChangeDetectionStrategy.OnPush'];
-      if (templateContent) {
+
+      if (fileContent.includes('lang="md"')) {
```
Comment (nartc) [inline packages/vite-plugin-angular/src/lib/authoring/analog.ts:82]: suggestion: I'd compute this flag ahead and check if this is true then skip the `TEMPLATE_TAG_REGEX.exec()` in this file.

**Comment by luishcastroc** [comment]
> Love love love this PR! Just so I understand what is going on. If you use the `<template lang="md">` tag, the template get's turned into a virtual file, which is recognized by it's "virtual" path and parsed through the `marked` service while it is read. Which means the template contains "regular" html. Importantly this "regular" html can contain Angular elements, because it will then be again parsed through the default `.analog` readFile function, which means we end up with markdown that supports Angular components. Is that about right? 😍 

Pretty much what I see as well

**Comment by goetzrobin** [comment]
> > Love love love this PR! Just so I understand what is going on. If you use the `<template lang="md">` tag, the template get's turned into a virtual file, which is recognized by it's "virtual" path and parsed through the `marked` service while it is read. Which means the template contains "regular" html. Importantly this "regular" html can contain Angular elements, because it will then be again parsed through the default `.analog` readFile function, which means we end up with markdown that supports Angular components. Is that about right? 😍
> 
> Yep, that's about right.

Love it! One thing that comes to mind as a potential future enhancement is allowing developers to pass their own custom marked instance through the vite config to the `MarkedSetupService`.

**Comment by brandonroberts** [comment]
> > > Love love love this PR! Just so I understand what is going on. If you use the `<template lang="md">` tag, the template get's turned into a virtual file, which is recognized by it's "virtual" path and parsed through the `marked` service while it is read. Which means the template contains "regular" html. Importantly this "regular" html can contain Angular elements, because it will then be again parsed through the default `.analog` readFile function, which means we end up with markdown that supports Angular components. Is that about right? 😍
> > 
> > 
> > Yep, that's about right.
> 
> Love it! One thing that comes to mind as a potential future enhancement is allowing developers to pass their own custom marked instance through the vite config to the `MarkedSetupService`.

Definitely! Developers may want custom highlighting also instead of Prism as mentioned by @nartc

File: packages/vite-plugin-angular/src/lib/authoring/analog.spec.ts [hunk 3]
```diff
@@ -109,6 +109,12 @@ effect(() => {
 </script>
 `;
 
+const MARKDOWN_CONTENT = `
+<template lang="md">
+  # Hello World
+</template>
+`;
+
 describe('authoring ng file', () => {
```
Comment (luishcastroc) [inline packages/vite-plugin-angular/src/lib/authoring/analog.spec.ts:118]: should we re-name this for consistency?

**Review by luishcastroc — APPROVED** [review]

**Comment by joshuamorony** [comment]
Absolutely loving seeing this work, thank you and I can't wait to get my hands on it.

Do you have any opinion on using remark/rehype for parsing instead of marked (not sure if there is some technical limitation here)? I have a custom remark plugin for example that relies on using the Markdown AST, and it would be nice to have access to that ecosystem of plugins too.

I imagine we could then allow people to pass in whatever plugins they like with something like:

```ts
provideContent(withMarkdownRenderer({
   remarkPlugins: [[preview, { percentage: 0.4 }], relativeImages, getHeadings, getWordCount]
   rehypePlugins: [rehypeSlug, rehypeAutolinkHeadings, rehypePrism],
}))
```

**Comment by brandonroberts** [comment]
> Absolutely loving seeing this work, thank you and I can't wait to get my hands on it.
> 
> Do you have any opinion on using remark/rehype for parsing instead of marked (not sure if there is some technical limitation here)? I have a custom remark plugin for example that relies on using the Markdown AST, and it would be nice to have access to that ecosystem of plugins too.
> 
> I imagine we could then allow people to pass in whatever plugins they like with something like:
> 
> ```ts
> provideContent(withMarkdownRenderer({
>    remarkPlugins: [[preview, { percentage: 0.4 }], relativeImages, getHeadings, getWordCount]
>    rehypePlugins: [rehypeSlug, rehypeAutolinkHeadings, rehypePrism],
> }))
> ```

I don't think there's a technical limitation here, but more of a default option. I think this would be more of documenting how to provide your own renderer with `remark` for example.

As long as you implement a service that implements the [ContentRenderer](https://github.com/analogjs/analog/blob/main/packages/content/src/lib/content-renderer.ts) abstract class and provide it, you can provide your own implementation.

**Review by nartc — APPROVED** [review]

**Pull request merged**